rfc: optimize pd creation (#969)
oneapi-src/oneDNN: rfcs/20210129-optimize-pd-creation/README.md

# Optimization of Primitive Descriptor Creation

## Motivation

There are three majors steps to create a oneDNN primitive:
1. Create operation descriptor
2. Create primitive descriptor
3. Create primitive

In frameworks, all the three steps are performed every time when a user executes

> an operation. The last two steps are expensive hence they can significantly

Reviewer: ```suggestion
an operation. Typically, the last two steps are the most expensive hence they significantly
```

affect performance of the operation.

> oneDNN recently introduced a cache for primitives to significantly reduce primitive

Reviewer: There are many many notions in this RFC which can make reading it a bit confusing. I would suggest to emphasize that the entity the RFC is based on is explicitly called `primitive cache`, and use it in as many places as possible.

Author: Thanks for suggesting a better writing for the places you find confusing, it definitely makes the RFC better.

creation time - the most expensive step. The second most expensive step is creating

> the primitive descriptor. The cost may very from several microsecond to hundreds

Reviewer: ```suggestion
the primitive descriptor. The cost may vary from several microseconds to hundreds
```

of microseconds depending on number of the primitive implementations and time of
primitive descriptor initialization.

> Such overhead is significant because the execution time can also fall in this range.

Reviewer: ```suggestion
Such overhead is significant because it is comparable to the execution time of the corresponding primitive.
```

> This is the motivation to start optimizing the primitive descriptor creation step.

Reviewer: I believe this sentence is redundant, the entire paragraph is about the motivation.

Author: I don't mind removing it if you think it's redundant.

## Proposal

> Introducing one more cache for primitive descriptors doesn't look appealing for

Reviewer 2: Do I understand it right that by `one more cache` you mostly mean a separate and dedicated key for primitive descriptors? Is it because a dedicated to pd key would require a dedicated search and a separate space to store them?

Author: One more cache is about one more cache entity in the library and all the consequences of that, which include increased memory consumption, new API, additional maintenance cost, etc...

The key can be generalized, this is the main goal of this RFC.

the following reasons:
1. Bad user experience because user has to manage two caches
2. Maintenance cost (maintain two multithreaded caches)
3. Increases memory consumption
4. New API to control one more cache

The proposal is to implement an internal mechanism that would significantly
reduce time of primitive descriptor creation but do not expose the mechanism to user.

> The idea is to use existing cache for primitives to implement the mechanism

Reviewer: ```suggestion
The idea is to use existing *primitive cache* to implement the mechanism
```

> because the cache already contains all the abstractions that are created during those

Reviewer: ```suggestion
because the cache already contains all the abstractions that are created during the above
```

3 steps. The `primitive_t` abstraction contains `primitive_desc_t` which contains
`op_desc_t` and `primitive_attr_t`.

> According to the primitive cache design there was added an ability to share

Reviewer: ```suggestion
According to the *primitive cache* design there was added an ability to share
```

> the `primitive_desc_t` abstraction with `shared_ptr` therefore we can simply use it

Reviewer: My overall suggestion is to use short sentences to simplify reading.

```suggestion
the `primitive_desc_t` abstraction with `shared_ptr`. Therefore we can simply use it
```

to create a new `primitive_desc_t` with minimum overhead.

> Since primitive descriptor and primitive are created for each execution of an

Reviewer: I would suggest to put this block at the beginning of this section, since it describes the motivation for pursuing this option. Or it can be dropped altogether.

Author: The idea about reusing primitive cache is mentioned in the previous paragraph first time hence this sentence seems to be in the right place.

operation in frameworks and the approach to caching primitives works it will
also work for primitive descriptors.

## Design

> Basically, the cache related work is done as part of enabling the primitive cache.

Reviewer: ```suggestion
Basically, the cache related work is done as part of enabling the *primitive cache*.
```

> The main goal now is to provide a bridge between API that is responsible for creating
> primitive descriptors and the cache for primitives.

Reviewer: ```suggestion
The main goal now is to provide a bridge primitive descriptors creation API and the *primitive cache*.
```

### Key for Primitive Descriptor

> The key which is used to find primitives in the cache can be used to find primitive
> descriptors as well however, there are a few problems to solve.

Reviewer: ```suggestion
The key which is used to find primitives in the *primitive cache* can be used to find primitive
descriptors too. However, there are a few problems to solve.
```

Currently, the key looks as follows:
```cpp
struct key_t {
     key_t(const primitive_desc_t *pd, const engine_t *engine, int impl_nthr);
    // Obtained from `primitive_desc_t`
    op_desc_t *op_desc_;
    primitive_attr_t *attr_;
    std::type_index impl_id_;
    std::vector<memory_desc_t> bwd_mds;

    // Obtained from `op_desc_t`
    primitive_kind_t primitive_kind_;

    // Obtained from `engine_t`
    engine_kind_t engine_kind_;
    runtime_kind_t runtime_kind_;
    device_id_t device_id_;

    // Obtained from ctor
    int impl_nthr_;

    // ... //
```

> When creating a key to find a primitive in the cache there is a primitive descriptor

Reviewer: ```suggestion
When creating a key to find a primitive in the *primitive cache* there is a primitive descriptor
```

which is used to obtain some of the required data. However, in case of primitive descriptor
creation for all primitives except reorder, concat and sum user creates `op_desc_t`,
`primitive_attr_t`, `engine_t` and optionally a hint primitive descriptor. Therefore
`impl_id_`, `bwd_mds` and `op_desc_t` for the aforementioned primitives cannot
be obtained at that point.

#### Implementation ID

The rationale behind having an implementation ID is to be able to differentiate
the same primitives which have different underlying implementations. Such primitives
can be created when user uses primitive descriptor iterator to manually iterate
through the list of implementations.

> The problem here is that the primitive descriptor is only going to be created and

Reviewer: I'm not sure I understand this sentence.

Author: What I was trying to tell here is that you have to know what implementation to look for to create a key.
The only way to get that information is to create the primitive descriptor for the given operation descriptor and attributes but this is what we are trying to avoid. Because of that, there is the idea about the new implementation ID which would help to "predict" implementation.

currently there is no way to get implementation ID without creating primitive
descriptor.

> Currently, `std::type_index` of a primitive descriptor of a particular implementation

Reviewer: Are you referring to `primitive_desc_t`?

I think it may also be worth introducing a small section (somewhere at the beginning of this RFC), where you specify how to distinguish between user-facing entities (dnnl::primitive{_desc}) and non-user-facing ones (primitive{_desc}_t). Otherwise I find it challenging to distinguish the two from the context.

In case you refer only to a single entity through this document, please specify it explicitly.

Author: No, I'm not referring to `primitive_desc_t`. The idea is to state class name explicitly when it's needed otherwise use the abstraction name.

This sentence is referring to `jit_avx512_common_convolution_fwd_t::pd_t`, `jit_avx2_convolution_fwd_t::pd_t`, etc. Since it's not about a particular class  the abstraction name is used.

Typically, the assumption for architecture related RFCs is that the reader is familiar with oneDNN architecture. Sometimes it's worth it to describe some additional details but in this particular case I don't think the description you are suggesting is appropriate because  it's not important in this context.

is used as the implementation ID. In general, this approach has only one minor issue,

> it doesn't make it possible to build oneDNN without RTTI support, though it's

Reviewer: Why is RTTI important?

Author: It doesn't say it is. Usually RTTI support is enabled by default. For some case it can be disabled to reduce binary size.

unclear whether such ability is needed. However, to be able to use the existing

> key and hence the existing cache for primitives to speed up primitive descriptor

Reviewer: ```suggestion
key and hence the existing *primitive cache* to speed up primitive descriptor
```

creation the implementation ID must be reworked.

The idea is to replace the existing implementation ID with a new one which is
defined as follows:

Implementation ID is a relative index of a valid implementation for the given
operation descriptor, attributes and a forward primitive descriptor hint.

For example there are operation descriptor and attributes for which some of
the implementations in the implementation list are valid. Then the implementation
IDs will be the following:

| Implementation ID | Implementation index in the list | Valid |
| ----------------- | -------------------------------- | ----- |
|       -           |                0                 |   -   |
|       0           |                1                 |   +   |
|       -           |                2                 |   -   |
|       1           |                3                 |   +   |
|       -           |                4                 |   -   |
|       2           |                5                 |   +   |

In other words, the implementation IDs are the indexes in a hypothetical list

> of valid implementations for the given operation descriptors and attributes.

Reviewer 3: I am not a huge fan of this approach since it ties the primitive cache to our dispatching mechanism.
If I understand correctly, if we want to change our dispatching mechanism, this would require to rework how the primitive (descriptor) cache works as well under the current proposal.

Author: It's not something that is only fair for this proposal. This is fair for whatever caching approach (primitive or pds) because it requires a key. This means that the cache has to be attached to something. Currently it's attached to operation descriptor, attributes, hint pd, device, number of threads, implementation ID, etc. If we want to change any of those we have to adjust the cache.

> The implementation ID can be non-zero only when using a primitive descriptor

Reviewer: How do you think we should handle the situation below (let's assume an empty cache for simplicity):
~~~cpp
auto pd = convolution_forward::primitive_desc(op_desc, attr, eng); // ----> cache_miss, ID = 0
pd.next_impl(); // ----> cache miss, ID = 1
auto prim = convolution_forward(pd); // ---> cache_miss, ID = 1, cache entry created
auto pd = convolution_forward::primitive_desc(op_desc, attr, eng); // ----> cache_miss or cache_hit?
~~~

In other words, should we extract the primitive descriptor implementation from the cache regardless of whether the user has iterated using `next_impl()`?

If we only extract the implementation ID that the user-facing `primitive_desc` object contains, then the execution will always have to iterate over the implementation list until finding the one available in the primitive cache.

Author: In the code snippet it will a cache miss because it will start with the first valid implementation but it was the second implementation that was cached.

Although, iterating through the list of implementations will not benefit much from approach described in this RFC it's not the common case we are trying to optimize. Basically, frameworks and other users do not use the iterator unless they need a workaround to fall back to a reference/simple implementation. But in this case the performance will suffer anyway.

iterator to iterate through the list of implementations. Each attempt to iterate
through the list to get the next valid implementation increments the implementation
ID. In all other cases the implementation ID will be 0.

Also, there are 3 primitives that do not support the iterator, those are sum,
concat and reorder, therefore the implementation ID for those primitives is always
0.

The implementation ID will be stored in `primitive_desc_t` so that it can be
obtained to create a key during primitive creation.

#### Automatic Algorithm Selection for Convolution

oneDNN supports `dnnl::algorithm::convolution_auto` algorithm that instructs
the library to automatically select the best algorithm based on the heuristics
that take into account tensor shapes and the number of logical processors

> available. The algorithm is specified with operation descriptor.

Reviewer: ```suggestion
available. The algorithm is specified for convolution operation descriptor.
```

oneDNN provides a query mechanism for getting information from oneDNN abstractions.
For example a user can query a primitive descriptor from created primitive. oneDNN
guarantees that the primitive descriptor that the user used to create the primitive
is equal to the the primitive descriptor that the query mechanism returns. This is
guaranteed not only for primitive descriptors but for all other abstractions.
However, there is an exception - operation descriptor for convolution. The thing
here is that if the user created primitive descriptor for convolution using
an operation descriptor with algorithm `convolution auto` the query mechanism
will return the operation descriptor with the algorithm that the library selected.
This happens because oneDNN modifies the operation descriptor when creating
primitive descriptor.
Given that, `op_desc_t` that a user uses to create primitive descriptor and
`op_desc_t` which is obtained from primitive descriptor to create a key to find

> the requested primitive in the cache may differ in case of convolution therefore

Reviewer: ```suggestion
the requested primitive in the *primitive cache* may differ in case of convolution therefore
```

> they cannot be compared as is.

Reviewer: Can we bypass some comparisons if the user has specified `convolution_auto`, so that we extract the supported primitive (descriptor) from the cache if it's available?

Author: This is the idea described below in this RFC.

The idea to rely on the implementation ID to distinguish the algorithm and ignore

> the field in operation descriptor during comparison.

Reviewer 3: This places a constraint on our dispatching: all algorithms shall be dispatched from the same list.
Are we ok with that? @kwiersch what do you think?

Reviewer 4: @mgouicem As far as I understand, there is currently no difference in implementation lists for `auto`, `direct` and `wino` algorithms of convolution. So from that perspective the constraint is already satisfied. Looking forward, I think it would be hard to disentangle `direct` and `wino` since we have `auto` to choose between the two, so I have no objections.

#### Forward Primitive Descriptor Hint

For some primitives the hint can affect equality of primitive descriptors.

> Currently, there are only two such primitives: pooling and shuffle. Depending on the

Reviewer 2: > Currently, there are only two such primitives: pooling and shuffle.

I was thinking about LRN as well. Could you double check it's not really affected, please?

Author: It's not affected.

struct dnnl_concat_desc_t {
    dnnl_primitive_kind_t primitive_kind;
    dnnl_memory_desc_t dst_md;
    dnnl_dim_t n;
    dnnl_dim_t concat_dimension;
    std::vector<dnnl_memory_desc_t> src_mds;
};

struct dnnl_sum_desc_t {
    dnnl_primitive_kind_t primitive_kind;
    dnnl_memory_desc_t dst_md;
    dnnl_dim_t n;
    std::vector<float> scales;
    std::vector<dnnl_memory_desc_t> src_mds;
};
```

These operation descriptors are created and filled in constructors of primitive descriptors
for the corresponding primitives.

The main problem with these operation descriptors is that they have to be created
to create a key to find primitive descriptor in the cache of primitives. It brings
unnecessary overhead.
Since the primitive descriptors take ownership of all the data that is used to create
the internal operation descriptors there is no need to copy the data. The idea
is to replace deep copy with shallow copy for all data fields which are POD or
class types. This would allow to create the operation descriptors for the key
without undesired overhead.

The updated operation descriptors can look as follows:

```cpp
struct dnnl_reorder_desc_t {
    dnnl_primitive_kind_t primitive_kind;

>     const dnnl_memory_desc_t *src_md;

Reviewer: Does it mean that we will rely on `src_md` being alive until (at least) primitive creation?

Author: The `dnnl_reorder_desc_t ` will stored in `reorder_pd_t` which takes ownership of all the data. The lifetime of these entities are the same.

In case of creating a key to find primitive descriptor the `src_md` should be alive until we create a primitive descriptor which is the case.
